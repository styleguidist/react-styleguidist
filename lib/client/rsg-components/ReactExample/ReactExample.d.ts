import { Component } from 'react';
import PropTypes from 'prop-types';
import { TransformOptions } from 'buble';
interface ReactExampleProps {
    code: string;
    evalInContext(code: string): () => any;
    onError(err: Error): void;
    compilerConfig?: TransformOptions;
}
export default class ReactExample extends Component<ReactExampleProps> {
    static propTypes: {
        code: PropTypes.Validator<string>;
        evalInContext: PropTypes.Validator<(...args: any[]) => any>;
        onError: PropTypes.Validator<(...args: any[]) => any>;
        compilerConfig: PropTypes.Requireable<object>;
    };
    shouldComponentUpdate(nextProps: ReactExampleProps): boolean;
    private getExampleComponent;
    render(): JSX.Element | null;
}
export {};
