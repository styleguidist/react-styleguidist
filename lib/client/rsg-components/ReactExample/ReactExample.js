import "core-js/modules/es.object.get-prototype-of";
import "core-js/modules/es.object.to-string";
import "core-js/modules/es.reflect.construct";
import "core-js/modules/es.regexp.to-string";

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Wrapper from 'rsg-components/Wrapper';
import compileCode from '../../utils/compileCode';
import splitExampleCode from '../../utils/splitExampleCode';
/* eslint-disable react/no-multi-comp */

var ReactExample = /*#__PURE__*/function (_Component) {
  _inheritsLoose(ReactExample, _Component);

  var _super = _createSuper(ReactExample);

  function ReactExample() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = ReactExample.prototype;

  _proto.shouldComponentUpdate = function shouldComponentUpdate(nextProps) {
    return this.props.code !== nextProps.code;
  } // Run example code and return the last top-level expression
  ;

  _proto.getExampleComponent = function getExampleComponent(compiledCode) {
    return this.props.evalInContext("\n\t\t\t" + compiledCode + "\n\t\t");
  };

  _proto.render = function render() {
    var _this$props = this.props,
        code = _this$props.code,
        _this$props$compilerC = _this$props.compilerConfig,
        compilerConfig = _this$props$compilerC === void 0 ? {} : _this$props$compilerC,
        onError = _this$props.onError;
    var compiledCode = compileCode(code, compilerConfig, onError);

    if (!compiledCode) {
      return null;
    }

    var _splitExampleCode = splitExampleCode(compiledCode),
        example = _splitExampleCode.example;

    var ExampleComponent = this.getExampleComponent(example);
    var wrappedComponent = /*#__PURE__*/React.createElement(Wrapper, {
      onError: onError
    }, /*#__PURE__*/React.createElement(ExampleComponent, null));
    return wrappedComponent;
  };

  return ReactExample;
}(Component);

_defineProperty(ReactExample, "propTypes", {
  code: PropTypes.string.isRequired,
  evalInContext: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired,
  compilerConfig: PropTypes.object
});

export { ReactExample as default };