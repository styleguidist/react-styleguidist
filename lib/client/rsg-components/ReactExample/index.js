export { default } from 'rsg-components/ReactExample/ReactExample';