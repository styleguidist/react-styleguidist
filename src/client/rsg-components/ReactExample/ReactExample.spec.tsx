import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import noop from 'lodash/noop';
import renderer from 'react-test-renderer';
import { createRenderer } from 'react-test-renderer/shallow';
import ReactExample from '.';

const evalInContext = (a: string): (() => any) =>
	// eslint-disable-next-line no-new-func
	new Function('require', 'const React = require("react");' + a).bind(null, require);

it('should render code', () => {
	const renderer = createRenderer();
	renderer.render(
		<ReactExample code={'<button>OK</button>'} evalInContext={evalInContext} onError={noop} />
	);

	expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('should wrap code in Fragment when it starts with <', () => {
	const actual = renderer.create(
		<div>
			<ReactExample code="<span /><span />" evalInContext={evalInContext} onError={noop} />
		</div>
	);

	expect(actual.toJSON()).toMatchSnapshot();
});

it('should handle errors', () => {
	const onError = jest.fn();

	const renderer = createRenderer();
	renderer.render(
		<ReactExample code={'<invalid code'} evalInContext={evalInContext} onError={onError} />
	);

	expect(onError).toHaveBeenCalledTimes(1);
});

it('should set initial state with hooks', () => {
	const code = `
const [count, setCount] = React.useState(0);
<button>{count}</button>
	`;
	const { getByRole } = render(
		<ReactExample code={code} evalInContext={evalInContext} onError={noop} />
	);

	expect(getByRole('button').textContent).toEqual('0');
});

it('should update state with hooks', () => {
	const code = `
const [count, setCount] = React.useState(0);
<button onClick={() => setCount(count+1)}>{count}</button>
	`;
	const { getByRole } = render(
		<ReactExample code={code} evalInContext={evalInContext} onError={noop} />
	);
	fireEvent.click(getByRole('button'));

	expect(getByRole('button').textContent).toEqual('1');
});
