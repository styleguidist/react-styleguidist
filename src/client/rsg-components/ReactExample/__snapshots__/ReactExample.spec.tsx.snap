// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render code 1`] = `
<Wrapper
  onError={[Function]}
>
  <bound anonymous />
</Wrapper>
`;

exports[`should wrap code in Fragment when it starts with < 1`] = `
<div>
  <span />
  <span />
</div>
`;
